From f16ad3909abdd9f8648eda38ebbdd27fde9ea534 Mon Sep 17 00:00:00 2001
From: Richard Guy Briggs <rgb@redhat.com>
Date: Fri, 19 Dec 2014 00:02:07 -0500
Subject: [powerpc] add little endian flag to syscall_get_arch()

Message-id: <6f2c2fe9ced54b2c2257edc7b4a3686dab557c26.1418931685.git.rgb@redhat.com>
Patchwork-id: 102327
O-Subject: [RHEL7 PATCH 2/2] powerpc: add little endian flag to syscall_get_arch()
Bugzilla: 1169461
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>
RH-Acked-by: Paul Moore <pmoore@redhat.com>
RH-Acked-by: Kyle McMartin <kmcmarti@redhat.com>

Since both ppc and ppc64 have LE variants which are now reported by uname, add
that flag (__AUDIT_ARCH_LE) to syscall_get_arch() and add AUDIT_ARCH_PPC64LE
variant.

Without this,  perf trace and auditctl fail.

Mainline kernel reports ppc64le (per a058801) but there is no matching
AUDIT_ARCH_PPC64LE.

Since 32-bit PPC LE is not supported by audit, don't advertise it in
AUDIT_ARCH_PPC* variants.

See:
 https://www.redhat.com/archives/linux-audit/2014-August/msg00082.html
 https://www.redhat.com/archives/linux-audit/2014-December/msg00004.html

Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=8404592
BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1169461
Initial testing by Jiri Jaburek <jjaburek@redhat.com> using:
  http://download.devel.redhat.com/brewroot/work/tasks/6808/8326808/kernel-3.10.0-211.el7.test.src.rpm
  rpmbuild --target ppc64le --rebuild kernel-3.10.0-211.el7.test.src.rpm
Upstream commit: mpe: 63f13448d81c910a284b096149411a719cbed501

Signed-off-by: Richard Guy Briggs <rgb@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/include/asm/syscall.h b/arch/powerpc/include/asm/syscall.h
index 9caf413..48b918c 100644
--- a/arch/powerpc/include/asm/syscall.h
+++ b/arch/powerpc/include/asm/syscall.h
@@ -90,6 +90,10 @@ static inline void syscall_set_arguments(struct task_struct *task,
 
 static inline int syscall_get_arch(void)
 {
- return is_32bit_task() ? AUDIT_ARCH_PPC : AUDIT_ARCH_PPC64;
+ int arch = is_32bit_task() ? AUDIT_ARCH_PPC : AUDIT_ARCH_PPC64;
+#ifdef __LITTLE_ENDIAN__
+ arch |= __AUDIT_ARCH_LE;
+#endif
+ return arch;
 }
 #endif /* _ASM_SYSCALL_H */
diff --git a/include/uapi/linux/audit.h b/include/uapi/linux/audit.h
index 1baec56..16754ea 100644
--- a/include/uapi/linux/audit.h
+++ b/include/uapi/linux/audit.h
@@ -346,7 +346,9 @@ enum {
 #define AUDIT_ARCH_PARISC (EM_PARISC)
 #define AUDIT_ARCH_PARISC64 (EM_PARISC|__AUDIT_ARCH_64BIT)
 #define AUDIT_ARCH_PPC  (EM_PPC)
+/* do not define AUDIT_ARCH_PPCLE since it is not supported by audit */
 #define AUDIT_ARCH_PPC64 (EM_PPC64|__AUDIT_ARCH_64BIT)
+#define AUDIT_ARCH_PPC64LE (EM_PPC64|__AUDIT_ARCH_64BIT|__AUDIT_ARCH_LE)
 #define AUDIT_ARCH_S390  (EM_S390)
 #define AUDIT_ARCH_S390X (EM_S390|__AUDIT_ARCH_64BIT)
 #define AUDIT_ARCH_SH  (EM_SH)
-- 
1.7.1